src/functional/program_processors/codegens/gtfn/gtfn_ir_to_gtfn_im_ir.py

```diff
@@ -0,0 +1,274 @@
+import dataclasses
+from typing import Any, Dict, Iterable, List, Optional, Union
+
+import eve
+from eve import NodeTranslator, NodeVisitor
+from eve.utils import UIDGenerator
+from functional.program_processors.codegens.gtfn import gtfn_ir, gtfn_ir_common
+from functional.program_processors.codegens.gtfn.gtfn_im_ir import (
+    AssignStmt,
+    Conditional,
+    ImperativeFunctionDefinition,
+    InitStmt,
+    ReturnStmt,
+    Stmt,
+)
+
+
+def _find_connectivity(reduce_args: Iterable[gtfn_ir_common.Expr], offset_provider):
+    connectivities = []
+    for arg in reduce_args:
+        if isinstance(arg, gtfn_ir.FunCall) and arg.fun == gtfn_ir_common.SymRef(id="shift"):
+            assert isinstance(arg.args[-1], gtfn_ir.OffsetLiteral), f"{arg.args}"
+            connectivities.append(offset_provider[arg.args[-1].value])
+
+    if not connectivities:
+        raise RuntimeError("Couldn't detect partial shift in any arguments of reduce.")
+
+    if len({(c.max_neighbors, c.has_skip_values) for c in connectivities}) != 1:
+        # The condition for this check is required but not sufficient: the actual neighbor tables could still be incompatible.
+        raise RuntimeError("Arguments to reduce have incompatible partial shifts.")
+    return connectivities[0]
```

Comment:
In principle, that idea seems very reasonable to me. We toyed around with this idea during a discussion, wanting to use `trace_shifts.py` for that purpose but iirc decided its not really fit for that purpose. I'm currently not sure how we would write that pass, except you mean by "propagate the info in lowering" to propagate it starting from field view.

Reply:
No, I meant we should just use the same pass to find the connectivity. But since they are not on the same IR, it's maybe difficult, therefore we could run it only on itir, but store the information and propagate it to gtfn_ir.